teps/0092-scheduling-timeout.md

```diff
@@ -17,361 +24,474 @@ authors:
-for the moment), and use `execution` as the name of the field that does the same thing as the existing field:
+```yaml
+spec:
+  timeouts:
+    scheduling: "5m"
+    execution: "10"
+    total: "20m"
+```
+Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.
 
 ```yaml
 spec:
   timeouts:
-    scheduling: "5m" # this is the new option
-    execution: "10m" # this option would replace the existing `timeout` field
+    scheduling: "10m"
+    execution: "75m"
+    total: "0m"
 ```
+Invalid: This is invalid because `scheduling` and `execution` timeout does not add up to `total` timeout.
 
-Here is a list of some possible names for this option, comment if you prefer any of these (or have other ideas):
+### What if the pod gets rescheduled?
 
-* `execution`
-* `run`
-* `completion`
+If the `Pod` gets rescheduled, the `scheduling` timer would not restart, i.e. it would still be considered already scheduled.
 
-## Test Plan
+As a side note, we already do not handle this situation well, see [pipelines#2813](https://github.com/tektoncd/pipeline/issues/2813).
 
-* Create an end to end tests that ultimately creates an unschedulable pod and verify the `scheduling` timeout is enforced
-* Create an end to end test (or reconciler test if you fake the times?) that uses both the `scheduling` timeout and the
-  `execution` timeout, and ultimately the execution times out (e.g. the `Task` just sleeps) - ensure the full time of
-  both timeouts has been counted, e.g. not just the `execution` timeout
+
+### PipelineRun
+
+#### TaskRunTemplate
+
+In [TEP-0119][tep-0119], we added `TaskRunTemplate` to support setting configuration for all `TaskRun` in `PipelineRun`.  This proposal adds a new timeout configuration within a `TaskRunTemplate` that allows users to specify `scheduling`, `execution` and `total` timeout for all `TaskRuns` executed in a given `PipelineRun`.
+
+```yaml=
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunTemplate:
+    timeouts:
+      scheduling: "10m"
+      execution: "30m"
+      total: "40m"
+```
+
+#### TaskRunSpec
+
+In [TEP-0119][tep-0119], we added `TaskRunSpec` to support setting configuration for a specific `TaskRun` in `PipelineRun`. This proposal adds a new timeout configuration within a `TaskRunSpec` that allows users to specify a `scheduling`, `execution` and `total` timeout for a specific `TaskRun` using `TaskRunSpecs`.
+
+```yaml=
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  generateName: clone-test-build-
+spec:
+  pipelineRef:
+    name: clone-test-build
+  taskRunSpecs:
+    - pipelineTaskName: clone
+      timeouts:
+        scheduling: "10m"
+        execution: "30m"
+        total: "40m"
+```
+
+All the `TaskRuns` would have the specified timeouts applied.
```

Comment:
In that example, only the `TaskRun` for the `clone` pipeline task would have these specified timeouts.